### PR TITLE
fix(#1217): check throwing tests in bad-test-name

### LINKED ISSUE #1217: `bad-test-name` never checks `-`-prefixed throwing tests, so `[] -> it-works` passes while `[] +> it-works` is flagged
## What happens

`bad-test-name.xsl` selects test objects with an ad-hoc `starts-with(@name, '+')`
instead of the shared `eo:test-name()` function:

https://github.com/objectionary/lints/blob/b004cdbe8ec110b776fbf49755aa9f2a7dfd2f17/src/main/resources/org/eolang/lints/tests/bad-test-name.xsl#L12-L16

```xml
<xsl:for-each select="/object//o[starts-with(@name, '+')]">
  <xsl:variable name="regexp" select="'^(can|cannot|accepts|rejects|stops-on)-'"/>
  <xsl:if test="not(matches(eo:escape-plus(@name), $regexp))">
```

The EO parser marks a truthy test attribute (`+> name`) with a `+` prefix and a
throwing test attribute (`-> name`) with a `-` prefix. `org/eolang/funcs/test-name.xsl`
states that "Both kinds are tests":

https://github.com/objectionary/lints/blob/b004cdbe8ec110b776fbf49755aa9f2a7dfd2f17/src/main/resources/org/eolang/funcs/test-name.xsl#L13-L16

```xml
<xsl:function name="eo:test-name" as="xs:boolean">
  <xsl:param name="name"/>
  <xsl:sequence select="starts-with($name, '+') or starts-with($name, '-')"/>
</xsl:function>
```

The motive of the lint says the rule applies to every unit test object:

> Every unit test object must be named to describe the behavior it verifies, starting
> with one of these prefixes: `can-`, `cannot-`, `accepts-`, `rejects-`, or `stops-on-`.

Because the `for-each` only sees `+`-prefixed names, a throwing test never reaches the
regexp and any name it carries is accepted. `stops-on-` in the allowed prefix list is
the idiomatic name for a throwing test, so the very objects that prefix targets are the
ones the rule skips.

`test-name.xsl` was introduced in f632e25 and used to replace `starts-with(@name, '+')`
in seven lints, including the sibling `misc/incorrect-test-object-name.xsl`.
`bad-test-name.xsl` was merged three hours earlier the same day (ca9cef9) and was not
part of that change, so it still carries the old check.

## Steps to reproduce

`truthy.eo`:

```eo
[] > foo
  [] > bar

  [] +> it-works
```

`throwing.eo`, identical except for the test marker:

```eo
[] > foo
  [] > bar

  [] -> it-works
```

Run `new Source(new EoSyntax(new InputOf(Paths.get(file))).parsed()).defects()` on each
and filter for `bad-test-name`.

The parser produces `name="+it-works"` for the first file and `name="-it-works"` for the
second; the objects are otherwise identical.

## Observed

`truthy.eo`:

```text
[bad-test-name/S WARNING]:4 The name of the test object "+it-works" must start with one of the prefixes "can-,⌴cannot-,⌴accepts-,⌴rejects-,⌴stops-on-"
```

`throwing.eo`: no `bad-test-name` defect at all.

## Expected behavior

`bad-test-name` should apply the prefix rule to throwing tests as well, so
`[] -> it-works` is reported just like `[] +> it-works`.

### PR DESCRIPTION
**Pull request by VasilevNStas**
Fixes #1217.

## Problem

`bad-test-name.xsl` selects test objects with an ad-hoc `starts-with(@name, '+')` instead of the shared `eo:test-name()` function. The EO parser marks a truthy test with a `+` prefix and a throwing test with a `-` prefix; the shared function (see `org/eolang/funcs/test-name.xsl`) treats both kinds as tests. As a result a throwing test like `[] -> it-works` never reached the regexp and any name it carried was accepted, even though the motive says the rule applies to every unit test object.

## Solution

Use `eo:test-name(@name)` in the `for-each` predicate and strip the single-char prefix with `substring(@name, 2)` before applying the regexp (same pattern as the sibling `misc/incorrect-test-object-name.xsl`).

## Changes

- `src/main/resources/org/eolang/lints/tests/bad-test-name.xsl` — import `funcs/test-name.xsl`, select via `eo:test-name(@name)`, match `substring(@name, 2)`.
- `src/test/resources/org/eolang/lints/packs/single/bad-test-name/allows-throwing-prefix.yaml` — new pack: `[] -> stops-on-error` produces no defects.
- `src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml` — new pack: `[] -> it-works` produces one defect.

## Tests

- `LtByXslTest` — 402 tests green (was 1 failure for the new throwing pack before the fix).
- `mvn -Pqulice install` — clean.

### CODE REVIEW

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/allows-throwing-prefix.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# SPDX-FileCopyrightText: Copyright (c) 2016-2026 Objectionary.com
+# SPDX-License-Identifier: MIT
+---
+sheets:
+  - /org/eolang/lints/tests/bad-test-name.xsl
+asserts:
```
Comment (volodya-lombrozo) [inline src/test/resources/org/eolang/lints/packs/single/bad-test-name/allows-throwing-prefix.yaml:6]: @VasilevNStas do we have `defects: 0` here?

File: src/main/resources/org/eolang/lints/tests/bad-test-name.xsl [hunk 2]
```diff
@@ -7,13 +7,14 @@
...
+  <xsl:import href="/org/eolang/funcs/test-name.xsl"/>
   <xsl:import href="/org/eolang/funcs/defect-context.xsl"/>
   <xsl:output encoding="UTF-8" method="xml"/>
   <xsl:template match="/">
     <defects>
-      <xsl:for-each select="/object//o[starts-with(@name, '+')]">
+      <xsl:for-each select="/object//o[eo:test-name(@name)]">
         <xsl:variable name="regexp" select="'^(can|cannot|accepts|rejects|stops-on)-'"/>
-        <xsl:if test="not(matches(eo:escape-plus(@name), $regexp))">
+        <xsl:if test="not(matches(substring(@name, 2), $regexp))">
```
Comment (volodya-lombrozo) [inline src/main/resources/org/eolang/lints/tests/bad-test-name.xsl]: @VasilevNStas It looks strange to me. Why exactly two chars?

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# SPDX-FileCopyrightText: Copyright (c) 2016-2026 Objectionary.com
+# SPDX-License-Identifier: MIT
+---
+sheets:
+  - /org/eolang/lints/tests/bad-test-name.xsl
+asserts:
+  - /defects[count(defect[@severity='warning'])=1]
```
Comment (volodya-lombrozo) [inline src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml:7]: @VasilevNStas What is the difference between these two tests?

**Comment by VasilevNStas** [comment]
@volodya-lombrozo plz review this

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/allows-throwing-prefix.yaml (hunk 1 above)
Comment (VasilevNStas) [reply src/test/resources/org/eolang/lints/packs/single/bad-test-name/allows-throwing-prefix.yaml:6]: Good question. Yes, we expect exactly `0` defects here. The object `[] -> stops-on-error` is a throwing test (its `@name` becomes `-stops-on-error` in XMIR). After stripping the leading `-`, the name is `stops-on-error`, which starts with the valid prefix `stops-on-`, so the lint stays silent. The assert `/defects[count(defect[@severity="warning"])=0]` verifies exactly that.

File: src/main/resources/org/eolang/lints/tests/bad-test-name.xsl (hunk 2 above)
Comment (VasilevNStas) [reply src/main/resources/org/eolang/lints/tests/bad-test-name.xsl]: Fair point. `substring(@name, 2)` does NOT cut two characters — in XPath positions are 1-based, so it cuts exactly ONE leading character, the test marker `+` or `-`. But the magic number was indeed hard to read, so I made it self-documenting: now the prefix is stripped with `replace(@name, '^[+-]', '')`, which reads as "remove the leading plus or minus". Same result, no guessing. The change is in the upcoming commit on this branch.

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml (hunk 3 above)
Comment (VasilevNStas) [reply src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml:7]: Both packs use a throwing test (`-` prefix in XMIR). The difference is the name: `stops-on-error` in the `allows` pack starts with the valid prefix `stops-on-`, so `0` defects are expected. `it-works` in the `catches` pack does not start with any of the allowed prefixes, so exactly `1` defect is expected. In other words, the two packs are a positive/negative pair proving the lint reacts only to the prefix, not to the fact that the object is a test.

**Comment by VasilevNStas** [comment]
@volodya-lombrozo please re-review
I`m fixed it

File: src/main/resources/org/eolang/lints/tests/bad-test-name.xsl [hunk 4]
```diff
@@ -7,13 +7,15 @@
   <xsl:import href="/org/eolang/parser/_funcs.xsl"/>
   <xsl:import href="/org/eolang/funcs/lineno.xsl"/>
   <xsl:import href="/org/eolang/funcs/escape.xsl"/>
+  <xsl:import href="/org/eolang/funcs/test-name.xsl"/>
   <xsl:import href="/org/eolang/funcs/defect-context.xsl"/>
   <xsl:output encoding="UTF-8" method="xml"/>
   <xsl:template match="/">
     <defects>
-      <xsl:for-each select="/object//o[starts-with(@name, '+')]">
+      <xsl:for-each select="/object//o[eo:test-name(@name)]">
```
Comment (volodya-lombrozo) [inline src/main/resources/org/eolang/lints/tests/bad-test-name.xsl:15]: @VasilevNStas shouldn't we use `eo:test-attr` here instead?

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml (hunk 3 above)
Comment (volodya-lombrozo) [reply src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml:7]: To my knowledge there is no semantical difference between `it-works` and `stops-on-error` names. Or I'm wrong?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml (hunk 3 above)
Comment (VasilevNStas) [reply src/test/resources/org/eolang/lints/packs/single/bad-test-name/catches-throwing-bad-name.yaml:7]: @volodya-lombrozo You are right, there is no *semantic* difference between `it-works` and `stops-on-error` — both are arbitrary names. The lint works purely syntactically: it requires the name to start with one of the allowed prefixes. `stops-on-error` starts with `stops-on-`, so it passes; `it-works` does not start with any allowed prefix, so it is flagged. That is why one is in the `allows` pack and the other in the `catches` pack — they are a positive/negative pair proving the lint reacts only to the prefix.

That said, I am closing this PR: the underlying issue #1217 ("bad-test-name never checks `-`-prefixed throwing tests") is already solved, and better, by the merged PR #1244 (`fix(#1243): enforce prefixes per test kind in bad-test-name`). It also switched to `eo:test-name()`, lints throwing tests, and goes further by separating the allowed prefixes per test kind (positive vs negative). Keeping this PR would only introduce a conflicting, inferior duplicate. Thank you for the review!

**Comment by VasilevNStas** [comment]
Closing this PR as it is a duplicate. Issue #1217 (throwing tests are never checked by `bad-test-name`) is already resolved — and resolved better — by the merged PR #1244 (fix(#1243): enforce prefixes per test kind in bad-test-name), which also lints `-`-prefixed tests via `eo:test-name()` and additionally separates the allowed prefixes per test kind. This branch would only add a conflicting, inferior duplicate. Thanks for the review @volodya-lombrozo.

**Comment by 0crat** [comment]
@volodya-lombrozo Thanks for the review! You've earned +4 points for this: +12 as a basis; -5 for very few (3) comments; -4 for too few (30) hits-of-code; +1 to give you at least something.  Your running score is [+547](https://www.eolang.org/zerocracy/objectionary-vitals.html); don't forget to check your [Zerocracy account](https://www.zerocracy.com/) too).
